File: go/1975.maximum-matrix-sum/solution.go
// Created by none at 2024/11/24 14:21
// leetgo: 1.4.9
// https://leetcode.com/problems/maximum-matrix-sum/

package main

import (
	"bufio"
	"fmt"
	"math"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func maxMatrixSum(matrix [][]int) int64 {
	ans := 0
	cnt := 0 // cnt neg
	has_zero := false
	y := math.MaxInt32
	for _, row := range matrix {
		for _, x := range row {
			if x > 0 {
				y = min(y, x)
				ans += x
			} else if x < 0 {
				y = min(y, -x)
				ans -= x
				cnt++
			} else {
				has_zero = true
			}
		}
	}

	if !has_zero && cnt&1 == 1 {
		ans -= 2 * y
	}
	return int64(ans)
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	matrix := Deserialize[[][]int](ReadLine(stdin))
	ans := maxMatrixSum(matrix)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/1975.maximum-matrix-sum/solution.rs
// Created by none at 2026/01/05 14:27
// leetgo: dev
// https://leetcode.com/problems/maximum-matrix-sum/

#![allow(unused_imports)]

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
use std::cmp::{Reverse, max, min};
use std::collections::*;
use std::i64::MAX;
use std::mem::swap;

impl Solution {
    pub fn max_matrix_sum(matrix: Vec<Vec<i32>>) -> i64 {
        // find the max neg and min pos
        // if count of neg is odd, then sub 2 * min(-neg, pos)

        let mut neg = i32::MIN;
        let mut pos = i32::MAX;
        let mut ans = 0;
        let mut cnt = 0;
        for row in matrix {
            for x in row {
                ans += x.abs() as i64;
                if x <= 0 {
                    cnt += 1;
                    neg = max(neg, x)
                } else {
                    pos = min(pos, x)
                }
            }
        }

        // println!("{neg} {pos} {cnt} ");
        if cnt & 1 == 1 {
            ans - 2 * min(-neg, pos) as i64
        } else {
            ans
        }
    }
}

// @lc code=end

fn main() -> Result<()> {
    let matrix: Vec<Vec<i32>> = deserialize(&read_line()?)?;
    let ans: i64 = Solution::max_matrix_sum(matrix).into();

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
